src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp

```diff
@@ -0,0 +1,997 @@
+  GEOS_MARK_FUNCTION;
+
+  mesh.getElemManager().forElementSubRegions( regionNames,
+                                              [&]( localIndex const,
+                                                   ElementSubRegionBase & subRegion )
+  {
+    // 2. Assume global component fractions have been prescribed.
+    // Initialize constitutive state to get fluid density.
+    updateFluidModel( subRegion );
+
+  } );
+  
+  // for some reason CUDA does not want the host_device lambda to be defined inside the generic lambda
+  // I need the exact type of the subRegion for updateSolidflowProperties to work well.
+  mesh.getElemManager().forElementSubRegions< CellElementSubRegion,
+                                              SurfaceElementSubRegion >( regionNames, [&]( localIndex const,
+                                                                                           auto & subRegion )
+  {
+    // 4. Initialize/update dependent state quantities
+
+    // 4.1 Update the constitutive models that only depend on
+    //      - the primary variables
+    //      - the fluid constitutive quantities (as they have already been updated)
+    // We postpone the other constitutive models for now
+    // In addition, to avoid multiplying permeability/porosity bay netToGross in the assembly kernel, we do it once and for all here
+    arrayView1d< real64 const > const netToGross = subRegion.template getField< fields::flow::netToGross >();
+    CoupledSolidBase const & porousSolid =
+      getConstitutiveModel< CoupledSolidBase >( subRegion, subRegion.template getReference< string >( viewKeyStruct::solidNamesString() ) );
+    PermeabilityBase const & permeabilityModel =
+      getConstitutiveModel< PermeabilityBase >( subRegion, subRegion.template getReference< string >( viewKeyStruct::permeabilityNamesString() ) );
+    permeabilityModel.scaleHorizontalPermeability( netToGross );
+    porousSolid.scaleReferencePorosity( netToGross );
+    saveConvergedState( subRegion ); // necessary for a meaningful porosity update in sequential schemes
+    updatePorosityAndPermeability( subRegion );
+
+    // Now, we initialize and update each constitutive model one by one
+
+    // 4.2 Save the computed porosity into the old porosity
+    //
+    // Note:
+    // - This must be called after updatePorosityAndPermeability
+    // - This step depends on porosity
+    string const & solidName = subRegion.template getReference< string >( viewKeyStruct::solidNamesString() );
+    CoupledSolidBase const & porousMaterial = getConstitutiveModel< CoupledSolidBase >( subRegion, solidName );
+    porousMaterial.initializeState();
+
+    // 4.3 Initialize/update the relative permeability model using the initial phase volume fraction
+    //     This is needed to handle relative permeability hysteresis
+    //     Also, initialize the fluid model
+    //
+    // Note:
+    // - This must be called after updatePhaseVolumeFraction
+    // - This step depends on phaseVolFraction
```

Comment:
I didn't see the phase volume fraction update. Or maybe, are saturations primary variables in this solver? I can't quite remember.